File: nexus-sdk/token_cache.go
package oauthsdk

import (
	"sync"
	"time"
)

// CachedToken holds a resolved token with its expiration metadata.
type CachedToken struct {
	AccessToken string
	TokenType   string
	ExpiresAt   time.Time

	// HeaderName is the HTTP header to set (e.g., "Authorization", "X-API-Key").
	// Defaults to "Authorization" for OAuth2 strategies.
	HeaderName string

	// ValuePrefix is prepended to the token value (e.g., "Bearer ").
	// Empty string for raw API-key strategies.
	ValuePrefix string
}

// TokenCache is a thread-safe, TTL-aware in-memory cache for resolved tokens.
// It is keyed by workspace+provider pairs and automatically evicts expired entries.
type TokenCache struct {
	mu      sync.RWMutex
	entries map[string]CachedToken
	buffer  time.Duration // safety buffer before actual expiry
}

// NewTokenCache creates a new TokenCache.
// buffer is the safety margin subtracted from the expiry time (e.g., 30s)
// so the caller re-fetches slightly before the token actually expires.
func NewTokenCache(buffer time.Duration) *TokenCache {
	if buffer <= 0 {
		buffer = 30 * time.Second
	}
	return &TokenCache{
		entries: make(map[string]CachedToken),
		buffer:  buffer,
	}
}

func cacheKey(workspaceID, provider string) string {
	return workspaceID + ":" + provider
}

// Get returns a cached token if it exists and has not expired.
// Returns nil if the token is missing or stale.
func (tc *TokenCache) Get(workspaceID, provider string) *CachedToken {
	tc.mu.RLock()
	entry, ok := tc.entries[cacheKey(workspaceID, provider)]
	tc.mu.RUnlock()
	if !ok {
		return nil
	}
	if time.Now().After(entry.ExpiresAt.Add(-tc.buffer)) {
		// Token is expired or within the safety buffer — evict and return nil.
		tc.Delete(workspaceID, provider)
		return nil
	}
	return &entry
}

// Set stores a token in the cache.
func (tc *TokenCache) Set(workspaceID, provider string, token CachedToken) {
	tc.mu.Lock()
	tc.entries[cacheKey(workspaceID, provider)] = token
	tc.mu.Unlock()
}

// Delete removes a cached token.
func (tc *TokenCache) Delete(workspaceID, provider string) {
	tc.mu.Lock()
	delete(tc.entries, cacheKey(workspaceID, provider))
	tc.mu.Unlock()
}

File: nexus-sdk-python/nexus_sdk/token_cache.py
"""Thread-safe, TTL-aware in-memory token cache."""

from __future__ import annotations

import threading
import time

from nexus_sdk.types import CachedToken


class TokenCache:
    """
    Thread-safe in-memory cache for resolved tokens.

    Keyed by workspace+provider pairs. Automatically evicts entries
    that are within the safety buffer of their expiration time.

    Args:
        buffer_seconds: Safety margin (in seconds) before the actual expiry
            time. The cache will treat the token as expired this many seconds
            early, forcing a re-fetch before the upstream token actually expires.
    """

    def __init__(self, buffer_seconds: float = 30.0):
        self._entries: dict[str, CachedToken] = {}
        self._lock = threading.RLock()
        self._buffer = buffer_seconds

    @staticmethod
    def _key(workspace_id: str, provider: str) -> str:
        return f"{workspace_id}:{provider}"

    def get(self, workspace_id: str, provider: str) -> CachedToken | None:
        """Return a cached token if it exists and is not expired."""
        key = self._key(workspace_id, provider)
        with self._lock:
            token = self._entries.get(key)
            if token is None:
                return None
            if time.time() > (token.expires_at - self._buffer):
                # Expired or within safety buffer — evict
                del self._entries[key]
                return None
            return token

    def set(self, workspace_id: str, provider: str, token: CachedToken) -> None:
        """Store a token in the cache."""
        key = self._key(workspace_id, provider)
        with self._lock:
            self._entries[key] = token

    def delete(self, workspace_id: str, provider: str) -> None:
        """Remove a cached token."""
        key = self._key(workspace_id, provider)
        with self._lock:
            self._entries.pop(key, None)
